Move post/page title to top toolbar

artemiomorales opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
This PR addresses #27093 and moves the post/page title to the top toolbar.

## Why?
Block themes and template editing give authors freedom on how to display titles of posts and pages on the frontend, and so, the previous implementation of the title — displaying it in the canvas — was increasingly unlikely to be an accurate representation of the frontend.

## How?
It largely duplicates the implementation of the toolbar from Full Site Editing mode, placing a button and dropdown into the header of the post and page edit window.

## Testing Instructions

Default view
1. Open a Post or Page.
2. See the title in the top toolbar
3. Click the dropdown button
4. Edit the title

"Show button text labels" view
1. Open a Post or Page
2. Open Gutenberg preferences
3. Enable "Show button text labels"
5. See updated dropdown button
6. Edit the title

## Screenshots or screencast <!-- if applicable -->
<img width="1688" alt="Screen Shot 2022-11-28 at 3 12 45 PM" src="https://user-images.githubusercontent.com/5360536/204386227-c22d799d-5c06-4e13-8e61-1bbd8369f75f.png">
<img width="1689" alt="Screen Shot 2022-11-28 at 4 35 39 PM" src="https://user-images.githubusercontent.com/5360536/204386249-55267af3-9a1f-4ec3-b6ec-30350a75afb0.png">

artemiomorales committed:
Move post/page title to top bar

The title's position in the editor can be
inaccurate depending on how themes are styled;
to more accurately represent how a post will
look when published, this commit moves the
title from the editor body to the top toolbar.

This mirrors the treatment for the toolbar in
Full Site Editing. For reference, see edit-site/
header-edit-mode/document-actions.

artemiomorales committed:
Prevent erroneous display of aria-label

When 'Show button text labels' is enabled in the
Gutenberg preferences, it causes the aria-label
to display beside the dropdown button unnecessarily.

This commit adds a style to prevent the aria-label
from appearing.

artemiomorales committed:
Add style variant to dropdown button

alexstine commented on packages/edit-post/src/components/header/header-toolbar/index.js:
```diff
@@ -14,10 +14,18 @@ import {
 } from '@wordpress/editor';
-import { Button, ToolbarItem } from '@wordpress/components';
-import { listView, plus } from '@wordpress/icons';
+import {
```
Is there a reason you are not using PostTitle from wordpress/editor in here?

artemiomorales replied:
PostTitle contains block editor logic that, as far as I can tell, is not necessary for the functioning of this feature. With that in mind, it seemed like a good opportunity to simplify the codebase by removing reference to and potentially deprecating PostTitle.

There may be onPaste or accessibility logic within PostTitle that could be of use here, but overall I figured this pull request could be the start of the conversation to see what might be the best way to approach, whether to revise PostTitle, remove it, or revise our approach to this feature altogether.

Am happy to hear additional thoughts!

alexstine reviewed (commented):
@artemiomorales Thanks for the PR. Sadly, this introduces a few major accessibility issues which I consider blockers.
1. "Edit document details" button should probably not be a sub menu in the event that it simply shows the title field.
2. The title field is rendered at the bottom of the DOM just below the block breadcrumbs. The title field needs to appear in relation to the button.
3. "Edit document details" does not say anything about editing the title. My problem is, we're going to start splitting hairs until no one knows how to find anything in the editor. From a general UX perspective, hiding everything in toggles can't be helping us long term.
4. Once you are done editing the title, there is no easy way back to the button that moved focus to the title field.
5. It is not clear to me how to save the title changes once done editing.

I'll flag this for additional reviews.

Thanks.

artemiomorales commented:
@alexstine Thanks so much for the review and for raising these concerns! I'll monitor the discussion on this pull request and #27093 and revise the code as the conversation develops.

artemiomorales committed:
Remove dropdown and convert title to input field

jameskoster commented:
Here's the current state of this PR:

<img width="1287" alt="Screenshot 2022-12-01 at 10 52 51" src="https://user-images.githubusercontent.com/846565/205034680-cfec90d7-13ac-4e24-9103-e47b71d4f7e3.png">

With the outline visible the input is too prominent, and the alignment looks strange. 

If we make the text blue (like all the other interactive text elements in the top bar) then hopefully we can remove the resting outline while maintaining clarity that the title is editable:

![title](https://user-images.githubusercontent.com/846565/205036589-bb5b8be3-e41b-4b95-bb85-9145dca59a86.gif)

If we include the tooltip, perhaps we don't need the outline at all. 

![title2](https://user-images.githubusercontent.com/846565/205037133-344b9209-359d-4588-b1ab-3d249cac3ae1.gif)

I think this is a bit cleaner, but appreciate the outline may be an a11y requirement.

alexstine commented on packages/edit-post/src/components/header/header-toolbar/index.js:
```diff
@@ -152,6 +157,15 @@ function HeaderToolbar() {
+					value={
+						title !== '' ? decodeEntities( title ) : 'Untitled'
+					}
+					aria-label={ __( 'Edit title' ) }
```
Instead of the `aria-label` here, what about this?

```suggestion
					label={ __( 'Edit title' ) }
```

If the design requires the label is still hidden, could pass `hideLabelFromVision={ true }`.

Also, the field is announced as an autocomplete. Maybe worth disabling that?

alexstine reviewed (commented):
@artemiomorales Accessibility is getting much closer. Thanks for your work on this.

artemiomorales committed:
Add tooltip, revise label for a11y, revise styles

artemiomorales commented:
@jameskoster @alexstine Great, thank you for the feedback! I went ahead and implemented James's suggestion with the gray background, added the tooltip, and revised the label to improve accessibility. Two notes: 

1. I made the input 98% of the center toolbar width, because otherwise, the design looks broken if a document has not been saved:
<img width="1362" alt="Screen Shot 2022-12-04 at 8 10 38 PM" src="https://user-images.githubusercontent.com/5360536/205529647-5563aa59-dc13-48ed-90c3-9f31b196fbee.png">

2. At certain resolutions, around 601-646px wide, the toolbar settings wrap and it doesn't look aesthetically pleasing. I could add `flex-wrap: nowrap` to the settings on the right to prevent this.
<img width="701" alt="Screen Shot 2022-12-04 at 8 19 03 PM" src="https://user-images.githubusercontent.com/5360536/205529995-b02199c4-88c6-4515-a9f8-3181335a737d.png">

Please let me know of any additional feedback or ideas 🙏

jameskoster committed:
Style adjustments

jameskoster commented:
Thanks for the updates :) 

I pushed some small style adjustments. 

> At certain resolutions, around 601-646px wide, the toolbar settings wrap and it doesn't look aesthetically pleasing

In the Site Editor, tools like undo/redo, the mode switcher, and the Inspector button are hidden at the $break-medium breakpoint. Might we try the same here?

artemiomorales commented:
@jameskoster The changes look great, thanks! However, titles now get cut off at around 27 characters. Maybe we can specify the width at different resolutions or find another solution to make sure more of the title is visible. When the title does get cut off, I think adding ellipses would be good, too. If that sounds good, I'll go ahead and implement those revisions.

<img width="1597" alt="Screen Shot 2022-12-05 at 9 58 50 AM" src="https://user-images.githubusercontent.com/5360536/205668943-9711cb6a-fd48-45ba-a990-94073de7d190.png">

> In the Site Editor, tools like undo/redo, the mode switcher, and the Inspector button are hidden at the $break-medium breakpoint. Might we try the same here?

I'm looking into this.

artemiomorales commented:
@jameskoster So at smaller resolutions, the undo/redo, mode switcher, and Inspector button are hidden, and their functionality is also disabled as a result. I don't think we can do the same with Save Draft, Preview, and Publish, which are key pieces of functionality.

One solution is we could hide the latter group behind an additional sidebar, though that seems like it would introduce additional complexity, in particular because the Publish button itself opens a sidebar, and we'd likely need a design for it. Another option: We could create an icon for the Preview button so that it takes up less space at smaller resolutions. Otherwise I think we can probably find a solution with manipulating widths and deactivating wrapping.

Which do you think would be the best approach?

jameskoster commented:
> Maybe we can specify the width at different resolutions or find another solution to make sure more of the title is visible. When the title does get cut off, I think adding ellipses would be good, too. If that sounds good, I'll go ahead and implement those revisions.

That seems fair, as long as the URL doesn't accommodate the entire space :)

> So at smaller resolutions, the undo/redo, mode switcher, and Inspector button are hidden, and their functionality is also disabled as a result. I don't think we can do the same with Save Draft, Preview, and Publish, which are key pieces of functionality.

I think if we hide the undo/redo, mode switcher, and Inspector button then there should be enough room? It's a bit cramped, but it does fit?

artemiomorales committed:
Fix wrapping bug and widen title input

Revised styles so that items in the header don't wrap, allowing
the left and center parts of the toolbar to shrink as needed.
We also now hide the undo, redo, mode switcher, and Inspector buttons
at resolutions below the medium breakpoint.

Also, I made it so more characters are visible in the title,
and added ellipsis when the title overflows.

artemiomorales commented:
@jameskoster Great, I doubled the width of the title and added some styles to keep that looking good at all resolutions.

I've also hidden the undo/redo, mode switcher, and Inspector buttons at resolutions below the medium breakpoint. Let me know what you think.

jameskoster commented:
Hmm, here's what I'm seeing now:

<img width="414" alt="Screenshot 2022-12-06 at 12 26 05" src="https://user-images.githubusercontent.com/846565/205912187-f4c11d2f-70c8-4dbd-8453-de9b6ed506e0.png">

Some of the tools are hidden on mobile, but I still see the Inspector button. The style changes seem to have reverted as well.

artemiomorales commented:
@jameskoster Did you clear your cache, and which browser are you using? I see the following on Chrome and Firefox, though did notice the buttons still wrap on Safari, which I can work to fix.

<img width="487" alt="Screen Shot 2022-12-06 at 9 37 52 AM" src="https://user-images.githubusercontent.com/5360536/205941598-abb3dfc4-00a6-49f3-9526-5bfcfe8749e0.png">

<img width="488" alt="Screen Shot 2022-12-06 at 9 49 23 AM" src="https://user-images.githubusercontent.com/5360536/205944095-753a0705-870b-43f2-8373-57012fc36169.png">

Also, I think I was unclear on the inspector button. Previously I thought we were referring to the Document Overview button, but are you referring to this button?

<img width="1178" alt="Screen Shot 2022-12-06 at 9 45 39 AM" src="https://user-images.githubusercontent.com/5360536/205943553-20683a4c-c29c-4bbd-8e40-b4a4563501b9.png">

jameskoster commented:
Yes, that one gets hidden in the Site Editor, so maybe it's okay to do the same here?

I've cleared cache and tried multiple browsers. Strange 🤔

artemiomorales committed:
Hide inspector button below medium breakpoint

artemiomorales commented:
@jameskoster 

> Yes, that one gets hidden in the Site Editor, so maybe it's okay to do the same here?

Ok got it, I just pushed a new commit to hide the button. By the way, we seem to be making relatively big changes to the post editor. Is there anyone else we should ping to make sure these changes would be good to go?

> I've cleared cache and tried multiple browsers. Strange 🤔

Hmm, just to double check: Have you pulled down the latest changes from the branch, and do you have the right branch `update/move-post-title-to-top-bar` checked out from origin https://github.com/artemiomorales/gutenberg-interactive-fiction-engine?

Note: I force pushed the branch, so you may need to delete the branch locally and pull down a new copy to see the changes. 

If everything I mentioned here is covered, maybe we can ask someone else to check and I can look deeper on my end to see if something is amiss there.

jameskoster commented:
> By the way, we seem to be making relatively big changes to the post editor. Is there anyone else we should ping to make sure these changes would be good to go?

Certainly!

> I force pushed the branch

I see all the latest changes (inspector button hidden on mobile), but the styles are still missing. Strange.

jasmussen commented:
Thanks for all the work and careful reviews here. Here's a GIF showing the post editor with the title moved to the block toolbar:

![gif](https://user-images.githubusercontent.com/1204802/206154296-41e35d33-371c-45cd-9d62-3ea86bc45d98.gif)

The input at the top works fairly well, it's minimal and does the job. This is also something that will benefit WYSIWYG in post designs where the title is omitted. 

However this comes with a few downsides, and sorry I haven't been able to look at this sooner. 

* Firstly, it's now somewhat less obvious how to add a title, by virtue both of this moving the cheese, but also by the footprint of the title being much smaller. 
* To an extent, it's a hit to the writing flow for folks who just want to knock out a quick blog post. I imagine they'd set focus in the title field in the canvas, add a title, press enter, then write the post. 
* It doesn't actually address the WYSIWYG aspect of the original issue, which perhaps isn't made too clear in the original issue, but is actually about making the title into a true _block_ which you can then include or exclude from your templates, and perhaps move inside a cover block if you like. 
* As the GIF shows, the block toolbar which normally shows up contextually above the selected block, it shows up _below_ the first paragraph since that's now close enough to the top of the canvas that it would otherwise be cropped if shown above. That's also a bit jarring for just writing.

In that light, as just a single PR, I don't think this one quite hits the mark. I think we need to pair this with the title being an actual block, and keeping it there by default — just with the option to remove it. What do you think?

jameskoster commented:
> I think we need to pair this with the title being an actual block, and keeping it there by default — just with the option to remove it. What do you think?

That doesn't seem like a change that is compatible with the post editor because it is essentially a zoomed-in edit view for the Post Content block. In terms of the template, the Post Title block will always reside _outside_ of Post Content.

The only way to do this would be to expose the full template in the post editor. But then you get into murky waters where folks assume that removing the Post Title will only affect a single post. I don't think we should go there.

The Post Title needs to be editable outside of the canvas to account for templates that do not include a Post Title block at all. It also needs to be removed from the canvas to preserve the WYSIWYG (see the frontend/backend comparison in https://github.com/WordPress/gutenberg/issues/27093).

jasmussen commented:
I'm not sure I see the post editor as zoomed in _post content_. Since title has always been a part of the post editor, I see it as _zoomed in title and post content_. If title was a block, you could do this:

<img src="https://user-images.githubusercontent.com/846565/200607467-2dbc2736-2033-47d2-b582-4cbc6787eaa0.png">

Insert a cover, move the title block in there. And of course you could remove it too, for when you want a title-less post. I worry shipping this as is, will be a worse blogging experience in the near term. #27093 for me feels about expressing the title more clearly, not just about moving the title to the top toolbar.

Mamaduka commented:
> To an extent, it's a hit to the writing flow for folks who just want to knock out a quick blog post. I imagine they'd set focus in the title field in the canvas, add a title, press enter, then write the post.

This is a great point. We even rely on similar flows in the e2e tests.

I think we should consider how much this UI change can affect our and users' writing flow.

Would it make sense to have a toolbar title option as a preference?

jasmussen commented:
To me it all goes back to the problem we're trying to solve. As I see it, the goal is to _enable more expression of the title in the post_, including the case where you want to remove the title from being displayed entirely. And for that case, there needs to be another affordance for editing the title.

aristath commented:
> To an extent, it's a hit to the writing flow for folks who just want to knock out a quick blog post. I imagine they'd set focus in the title field in the canvas, add a title, press enter, then write the post.

💯 That's what I always do when writing posts

aristath commented:
> To me it all goes back to the problem we're trying to solve. As I see it, the goal is to enable more expression of the title in the post, including the case where you want to remove the title from being displayed entirely. And for that case, there needs to be another affordance for editing the title.

I agree that we need to find a way to do that. However, perhaps that is something that should be done on a template-level and not on a post-level? (in which case it's pretty much what we already have - but we could streamline it a bit) 🤔

jameskoster commented:
> enable more expression of the title in the post

This is a template-level action, though. It doesn't make sense to do it on-canvas in the post editor for the reasons I mentioned above.

fabiankaegy commented:
I find myself agreeing with both parties here in this discussion. 

I also see the post editor as a Zoomed in view of the post content block since that is how it works when you create a template in the site editor. You can place the title completely independent from the post content. So placing the title again as a block inside the post content would be redundant and confusing.

But on the other hand the proposed spot here does make it less obvious especially since the placement of the title is a constant since the first versions of WordPress. 

On some builds I've recently been using a bit of custom CSS to place the title at the top of the post editor canvas but clearly differentiated from the canvas itself: 

![CleanShot 2022-12-07 at 12 36 43@2x](https://user-images.githubusercontent.com/20684594/206169640-6b408e7a-9f00-4217-b9fc-dd78dc53be6f.png)



This kind of solves both issues for me conceptually.

jameskoster commented:
> On some builds I've recently been using a bit of custom CSS to place the title at the top of the post editor canvas but clearly differentiated from the canvas itself

Yup, something like that could work. We had a similar concept way back when in the original issue https://github.com/WordPress/gutenberg/issues/27093#issuecomment-829078638

aristath commented:
> On some builds I've recently been using a bit of custom CSS to place the title at the top of the post editor canvas but clearly differentiated from the canvas itself:

That could do the trick, without breaking our writing flow... 👍

jasmussen commented:
I'm not convinced in more toolbars for this. Sorry to be the difficult/downer one here, but to me I see no benefits to the approaches outlined:

By moving it to the toolbar presumably the intent is to embrace WYSIWYG in cases where a post template doesn't have a post title, or has it elsewhere. Arguably those are 20% of cases, and as a result, the remaining 80% cases now have less WYSIWYG and a worse writing flow.

What is the main problem about a movable/deletable Title block, outside of this being technically challenging? To me that still seems like the way to thread the needle: bring about flexibility in the appearance, keep a good writing flow, and have an affordance for the title when the block is removed. The fact that `the_title` is technically separate from `the_content` seems like an inconsistency that can be explored at a later point if need be. CC: @mtias in case I'm missing some obvious context here.

fabiankaegy commented:
@jasmussen To be clear, I also don't think that this is the preferred end goal. From my perspective the approach outlined in https://github.com/WordPress/gutenberg/issues/27847 is really what we should thrive for here. 

In order to really be WYSIWYG the pos editor should show the title the same way it gets displayed by the template. That may just as well be as it is shown currently. But if the template moves inside a cover with the featured image underneath that should also display.

jasmussen commented:
#27847 is cool, but I think what I'm saying is that I wish there was a smaller step towards it than full on template view.

jameskoster commented:
> What is the main problem about a movable/deletable Title block

The problem is that the Post Title block is part of the template, not the post. If you put a Post Title inside the Post Content then you risk ending up with duplicate Post Titles on the frontend which is confusing and undesirable.

To be clear, this issue and PR _is not about the appearance of the post title_. 99.9% of the time that is a template-centric task. It's about manipulating the post title meta in a way that is wysiwyg compatible. The problem with the current implementation is that it _suggests_ wysiwyg, but the end result can be completely different.

It is a small problem in the grand scheme of things though, so perhaps it's not worth the hassle of changing anything here, and instead waiting for https://github.com/WordPress/gutenberg/issues/44461 or https://github.com/WordPress/gutenberg/issues/27847 as @fabiankaegy suggested.

hypest commented:
CC @WordPress/native-mobile as I think I remember some talks for moving the post title to the title bar in the native apps as well.

richtabor commented:
> By moving it to the toolbar presumably the intent is to embrace WYSIWYG in cases where a post template doesn't have a post title, has it elsewhere. Arguably those are 20% of cases, and as a result, the remaining 80% cases now have less WYSIWYG and a worse writing flow.

For posts, yea - but for pages I would think this could be flipped. An about page doesn’t have to have an “About” title on the page, whereas a post will have the post title somewhere.

alexstine commented:
@artemiomorales Still thinking you should pass `autocomplete="false"` to avoid having it read as an auto fill field.

Other than that, accessibility looks good.

Thanks.

alexstine commented:
Also, I saw it mentioned a couple times through the discussion. I am not sure we should ever find ourselves in a situation where a post or page does not have a title. That could throw off the heading order for accessibility and I can't imagine a page/post being useful at all without a title. If I can't find a title on a post or page on another site, I leave. I am not sure how the classic editor enforces this if it does at all, but I am not sure we should be making it clear to users that no title could work out for them.

jameskoster commented:
@alexstine some [post formats](https://wordpress.org/support/article/post-formats/) may not include titles. I think that's the main reason to consider the title-less scenario.

draganescu commented:
> The problem is that the Post Title block is part of the template, not the post. If you put a Post Title inside the Post Content then you risk ending up with duplicate Post Titles on the frontend which is confusing and undesirable.

This sounds like a limitation that should be overcome via improving the post title block and how the post editor handles it when it's optional/ removed. Moving the title to the title bar, is not bad per se but, to me, it does not solve the problem.

It feels weird to write posts without a visible title that looks and feels like a title. While _technically it is meta_, to many people I speculate the title is not meta :) but content.

The post title should be a block that exists in the post editor, if the post type has a title. Using the title bar, in my subjective perspective, makes that the post editor feel more like an ERP form than a composition canvas.

jameskoster commented:
Right, but exposing the Title without the rest of the template is also strange, because the appearance on the frontend may not match the on-canvas suggestion, accounting for position, widths, etc etc. 

It's not really feasible to say "we'll just display the Post Title and Post Content portions of the template in the post editor" because the two blocks may not be located anywhere near one another. Also where do you draw the line, should we include post meta blocks as well? 

If we try to do this then the lines blur very quickly and that is something we need to avoid lest we make the content/template boundaries even more confusing. For example if you style the Post Title block in one post it seems reasonable to expect those styles to be applied just to that post, but of course is not what happens, they're applied to the template which will affect multiple (all) posts.

We really need to think of the post editor as a zoomed-in view of the Post Content block imo. If you want to edit the appearance of the post title then you need to zoom out to view the full template. This is where we need to get the transition between post editor / site editor just right.

artemiomorales commented:
> It is a small problem in the grand scheme of things though, so perhaps it's not worth the hassle of changing anything here, and instead waiting for https://github.com/WordPress/gutenberg/issues/44461 or https://github.com/WordPress/gutenberg/issues/27847 as @fabiankaegy suggested.

Since it looks like there's a broader fix in the works, perhaps it's best to deprioritize this ticket to avoid introducing a new experience that will end up being changed anyway, which overall may just add confusion and cause friction for users as paradigms change.

I'll go ahead and close this pull request for now, though am happy to hear additional discussion either here or on the [original issue](https://github.com/WordPress/gutenberg/issues/27093).

draganescu commented:
> We really need to think of the post editor as a zoomed-in view of the Post Content block imo.

I agree. But in that case we should treat the post title like some item in the post summary tab of the inspector and maybe introduce some default behavior like 1st h1 is the title or something? Moving to the top bar does not cater enough :) IMO.

Sometimes the title in the list should be different than the title of the story. Plus the titles could be long on purpose. 

So I believe the perfectly correct statement at the top should be tackled in more depth.
